### ISSUE TITLE
Investigate HDT+ keying mechanisms

### ISSUE THREAD

**Issue by MatthewTingum** [Necrosys/x86-JTAG-Information#6]
Some people I've talked to suggest that modern AMD processors require a key to talk HDT
- Is this true?
- Can voltage glitching be used to get around this?
  - Which tools and timings have worked?
- If this is certificate based, how are dates validated?

**Comment by Necrosys** [comment]
> Some people I've talked to suggest that modern AMD processors require a key to talk HDT
> 
> * Is this true?
Yes or No. AFAIK, if it's already unlocked, you don't have to unlock.
Related patent: https://patents.google.com/patent/US20200210600A1/

> * Can voltage glitching be used to get around this?
>   
>   * Which tools and timings have worked?
Maybe possible? I think it will be big project.
Related research: https://github.com/PSPReverse/amd-sp-glitch/

> * If this is certificate based, how are dates validated?
To be researched.

**Comment by MatthewTingum** [comment]
> Yes or No. AFAIK, if it's already unlocked, you don't have to unlock.

Interesting! This is something to be looked in to. I had a short talk with the author of [this](http://www.loper-os.org/?p=1821) post and he thinks that the protocol was locked down sometime after 2012. He uses some lingo I don't quite understand or am unfamiliar with . i.e. "pill" means exploit or

**Comment by MatthewTingum** [comment]
Accidental close. Re-opened.

**Comment by Necrosys** [comment]
More related patents:
https://patents.google.com/patent/US10895597B2 (2018)
https://patents.google.com/patent/US9224012B2 (2013)

Maybe he meant JTAG password protection.

**Comment by MatthewTingum** [comment]
Awesome! I'll look into those docs. I've been invested in AMD debug protocol for some time. I wouldn't be surprised if certificates were used over passwords. There's little documentation but the Xbox One hacking scene has substantial knowledge of the platform security processor.

Certificates are used to determine if a development box is authorized AFAIK.

Bought a Sun Fire X4100 M2. These things almost certainly use a service processor to speak HDT using JTAG over MMIO. I think the version I bought doesn't have this capability. I'll probably get stuck buying another unit but that's life.

**Comment by MatthewTingum** [comment]
I think HDP and the link you've provided collide. From what I can tell, no one has tried to glitch with the intent of opening up HDP. Attackers thus far have wider dreams.

**Comment by MatthewTingum** [comment]
Obviously, the patent links you've provided are directly related. I was referring to that github repo.

**Comment by Necrosys** [comment]
> Bought a Sun Fire X4100 M2. These things almost certainly use a service processor to speak HDT using JTAG over MMIO. I think the version I bought doesn't have this capability. I'll probably get stuck buying another unit but that's life.

I just bought a Sun Fire X4100 M2, too. Nice machine. :)
![SunX4100M2_sunservice](https://github.com/Necrosys/x86-JTAG-Information/assets/19776699/5747aeda-67b5-4f4b-a651-cd645b6523a9)

**Comment by MatthewTingum** [comment]
Yeah, it is! I've been slowly working on recreating the portion of `hdtl` that lets you load instructions into icache and execute them on a CPU reset. By recreating, I mean making a PPC binary that will only run on this machine. I'm probably wasting my time with this step because we can learn what's said over JTAG without any reversing. I think it'll be a nice debugging supplement when it's done though.

Use the undocumented `-v` option of `hdtl` to trace JTAG. It was undocumented on version anyway...

**Comment by MatthewTingum** [comment]
Weirdly I haven't been able to reproduce my (verbose) output. I might have made a patch in real-time.
I wouldn't trust a stranger's docker image image but I've been using this compile static binaries: https://github.com/MatthewTingum/powerpc-gdb-static through which I've produced a static [gdbserver](https://github.com/MatthewTingum/powerpc-gdb-static)

I managed to get log messages like this:
```
/*
 * This is a rough trace of what it looks like to start speaking JTAG on the X4100
 *
 * hdt_jtag_init: non-G4 platform
 * hdt_signal_enable 01
 * IR bits:  5, IR width: 5, Num CPUs: 1, Num Cores: 2
 * found 1 cpus
 * hdt_signal_enable 00
 * hdtDiag: 1 CPU detected in JTAG chain
 * enable_hdt cpu 00
 * hdt_signal_enable 01
 * hdt_signal_enable 01
 * get_jtag_id: cpu 00 
 * num_cpus  01  ir_width 05  IR:bits: 04  data :  0 1 0 0 0
 *
 * NOTE: IDCODE is 00010
 * 
 * DR out :  1 1  0 0 0 0  0 0 0 0 0 0 1 1  0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 1 0 
 *  50413003  Part : 0413 , Version : 0005
 *
 * NOTE: I think our actual CPUID is 40f13 -- This might be a shortened version?
 *
 * hdtDiag: assert_dbreq: cpu0     01
 * hdtDiag: DBRDY set !!  0   
 * hdtDiag: assert_dbreq: cpu0     00
 * init_hdt enter cpu 00
 * num_cpus  01  ir_width 05  IR:bits: 04  data :  0 1 1 0 0 
 *
 * NOTE: USEHDT is 00110
 *
 * hdt_exe: cpu 00 core 00 cmd 02   data 00000000    size 04
 * DR in  :  0 1  0 1 0 0  0 0 0 0 1 0 0 0  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 *                                          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 * DR out :  1 0  0 0 0 0  0 0 0 0 1 0 0 0  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 *                                          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 * hdt_exe: exiting:     data 00000000 00000000
 * hdt_sys_ident: enter

```
Juicy shit. LFGO.

**Comment by Necrosys** [comment]
> ```
>  * NOTE: I think our actual CPUID is 40f13 -- This might be a shortened version?
>  ```
Yes, you are right. JTAG IDCODE for Opteron 22xx (Santa Rosa) is 0413.

**Comment by MatthewTingum** [comment]
Some of my recent findings indicate that more modern EPYC processors might not lock down HDT.

It seems that `IDCODE` is issued before `USEHDT` on the x4100 M2. This indicates to me that an `IDCODE` could be issued on even a platform that is "locked down".
